File: rubygems/pray-cli/lib/pray/ssh_agent.rb
# frozen_string_literal: true

require "socket"
require "base64"

module Pray
  module SshAgent
    module_function

    def sign(public_key, message)
      agent_socket = ENV["SSH_AUTH_SOCK"]
      raise Error.unsupported("SSH_AUTH_SOCK is not set") if agent_socket.to_s.empty?

      raw_key_bytes = parse_ssh_ed25519_public_key(public_key)
      public_key_blob = +""
      write_ssh_string(public_key_blob, "ssh-ed25519")
      write_ssh_string(public_key_blob, raw_key_bytes)

      payload = +""
      write_ssh_string(payload, public_key_blob)
      write_ssh_string(payload, message)
      write_u32(payload, 0)

      UNIXSocket.open(agent_socket) do |stream|
        write_ssh_message(stream, 13, payload)
        message_type, response = read_ssh_message(stream)
        unless message_type == 14
          raise Error.resolution("ssh agent returned unexpected message type: #{message_type}")
        end

        signature_blob = read_ssh_string_bytes(response)
        parse_ssh_signature_blob(signature_blob)
      end
    end

    def parse_ssh_ed25519_public_key(public_key)
      fields = public_key.split(/\s+/)
      algorithm = fields[0]
      raise Error.unsupported("public key must include an algorithm") unless algorithm
      unless algorithm == "ssh-ed25519"
        raise Error.unsupported("unsupported public key algorithm: #{algorithm}")
      end

      key_value = fields[1]
      raise Error.unsupported("public key must include key bytes") unless key_value

      blob = Base64.decode64(key_value)
      cursor = blob.b
      blob_algorithm = read_ssh_string!(cursor)
      unless blob_algorithm == "ssh-ed25519"
        raise Error.parse("public key", "ed25519 public key blob must start with ssh-ed25519")
      end

      key_bytes = read_ssh_string!(cursor)
      unless key_bytes.bytesize == 32
        raise Error.parse("public key", "ed25519 public key must be 32 bytes")
      end

      key_bytes
    end

    def parse_ssh_signature_blob(signature_blob)
      cursor = signature_blob.b
      algorithm = read_ssh_string!(cursor)
      unless algorithm == "ssh-ed25519"
        raise Error.unsupported("unsupported ssh signature algorithm: #{algorithm}")
      end

      Base64.strict_encode64(read_ssh_string!(cursor))
    end

    def write_ssh_message(stream, message_type, payload)
      buffer = message_type.chr + payload
      stream.write([buffer.bytesize].pack("N"))
      stream.write(buffer)
    end

    def read_ssh_message(stream)
      length = stream.read(4)&.unpack1("N")
      raise Error.resolution("empty ssh agent response") unless length

      buffer = stream.read(length)
      raise Error.resolution("truncated ssh agent response") if buffer.nil? || buffer.bytesize < 1

      [buffer.getbyte(0), buffer.byteslice(1, buffer.bytesize - 1).to_s]
    end

    def write_ssh_string(buffer, bytes)
      bytes = bytes.b
      buffer << [bytes.bytesize].pack("N")
      buffer << bytes
    end

    def write_u32(buffer, value)
      buffer << [value].pack("N")
    end

    def read_ssh_string!(cursor)
      raise Error.resolution("truncated ssh agent response") if cursor.bytesize < 4

      length = cursor.byteslice(0, 4).unpack1("N")
      cursor.slice!(0, 4)
      raise Error.resolution("truncated ssh agent response") if cursor.bytesize < length

      value = cursor.byteslice(0, length)
      cursor.slice!(0, length)
      value
    end

    def read_ssh_string_bytes(buffer)
      cursor = buffer.b
      read_ssh_string!(cursor)
    end
  end
end

File: npmjs/pray-cli/src/auth/ssh-agent.ts
import { readFileSync } from "node:fs";
import { connect } from "node:net";
import { PrayError } from "../errors.js";
import { httpPostJson, joinUrl } from "../http/client.js";
import { persistSession, type SessionFile } from "./session.js";

interface SshKeyChallengeResponse {
  challenge_id: string;
  challenge: string;
  fingerprint: string;
}

interface SshKeyLoginResponse {
  email: string;
  token: string;
}

export async function loginWithSshAgent(
  serverUrl: string,
  publicKeyPath: string,
  sessionRoot: string,
): Promise<SessionFile> {
  const publicKey = readFileSync(publicKeyPath, "utf8").trim();
  const base = trimTrailingSlash(serverUrl);
  const challenge = await postJson<SshKeyChallengeResponse>(
    joinUrl(base, "v1/auth/ssh-keys/challenge"),
    { public_key: publicKey },
  );
  const signature = await sshAgentSign(
    publicKey,
    Buffer.from(challenge.challenge, "utf8"),
  );
  const response = await postJson<SshKeyLoginResponse>(
    joinUrl(base, "v1/auth/ssh-keys/login"),
    {
      public_key: publicKey,
      challenge_id: challenge.challenge_id,
      signature,
    },
  );
  return persistSession(sessionRoot, {
    server_url: serverUrl,
    email: response.email,
    token: response.token,
    kind: "ssh_key",
    signer_fingerprint: challenge.fingerprint,
  });
}

async function sshAgentSign(
  publicKey: string,
  message: Buffer,
): Promise<string> {
  const agentSocket = process.env.SSH_AUTH_SOCK;
  if (!agentSocket) {
    throw PrayError.unsupported("SSH_AUTH_SOCK is not set");
  }
  const rawKeyBytes = parseSshEd25519PublicKey(publicKey);
  const publicKeyBlob = Buffer.concat([
    sshString(Buffer.from("ssh-ed25519")),
    sshString(rawKeyBytes),
  ]);
  const payload = Buffer.concat([
    sshString(publicKeyBlob),
    sshString(message),
    Buffer.from([0, 0, 0, 0]),
  ]);
  const response = await agentRequest(agentSocket, 13, payload);
  if (response.messageType !== 14) {
    throw PrayError.resolution(
      `ssh agent returned unexpected message type: ${response.messageType}`,
    );
  }
  const signatureBlob = readSshString(response.payload, 0).value;
  return parseSshSignatureBlob(signatureBlob);
}

function parseSshEd25519PublicKey(publicKey: string): Buffer {
  const fields = publicKey.split(/\s+/);
  const algorithm = fields[0];
  const keyValue = fields[1];
  if (!algorithm || !keyValue) {
    throw PrayError.unsupported("public key must include an algorithm and key");
  }
  if (algorithm !== "ssh-ed25519") {
    throw PrayError.unsupported(
      `unsupported public key algorithm: ${algorithm}`,
    );
  }
  const blob = Buffer.from(keyValue, "base64");
  const algorithmField = readSshString(blob, 0);
  if (algorithmField.value.toString("utf8") !== "ssh-ed25519") {
    throw PrayError.parse(
      "public key",
      "ed25519 public key blob must start with ssh-ed25519",
    );
  }
  const keyBytes = readSshString(blob, algorithmField.nextOffset).value;
  if (keyBytes.length !== 32) {
    throw PrayError.parse("public key", "ed25519 public key must be 32 bytes");
  }
  return keyBytes;
}

function parseSshSignatureBlob(signatureBlob: Buffer): string {
  const algorithm = readSshString(signatureBlob, 0);
  if (algorithm.value.toString("utf8") !== "ssh-ed25519") {
    throw PrayError.unsupported(
      `unsupported ssh signature algorithm: ${algorithm.value.toString("utf8")}`,
    );
  }
  const signature = readSshString(signatureBlob, algorithm.nextOffset).value;
  return signature.toString("base64");
}

function agentRequest(
  socketPath: string,
  messageType: number,
  payload: Buffer,
): Promise<{ messageType: number; payload: Buffer }> {
  return new Promise((resolve, reject) => {
    const socket = connect(socketPath);
    const chunks: Buffer[] = [];
    socket.on("error", (error) => reject(error));
    socket.on("data", (chunk) => chunks.push(chunk));
    socket.on("end", () => {
      try {
        resolve(parseAgentResponse(Buffer.concat(chunks)));
      } catch (error) {
        reject(error);
      }
    });
    const body = Buffer.concat([Buffer.from([messageType]), payload]);
    const frame = Buffer.alloc(4 + body.length);
    frame.writeUInt32BE(body.length, 0);
    body.copy(frame, 4);
    socket.end(frame);
  });
}

function parseAgentResponse(buffer: Buffer): {
  messageType: number;
  payload: Buffer;
} {
  if (buffer.length < 5) {
    throw PrayError.resolution("empty ssh agent response");
  }
  const length = buffer.readUInt32BE(0);
  const body = buffer.subarray(4, 4 + length);
  return { messageType: body[0]!, payload: body.subarray(1) };
}

function sshString(bytes: Buffer): Buffer {
  const buffer = Buffer.alloc(4 + bytes.length);
  buffer.writeUInt32BE(bytes.length, 0);
  bytes.copy(buffer, 4);
  return buffer;
}

function readSshString(
  buffer: Buffer,
  offset: number,
): { value: Buffer; nextOffset: number } {
  if (buffer.length < offset + 4) {
    throw PrayError.resolution("truncated ssh field");
  }
  const length = buffer.readUInt32BE(offset);
  const start = offset + 4;
  const end = start + length;
  if (buffer.length < end) {
    throw PrayError.resolution("truncated ssh agent response");
  }
  return { value: buffer.subarray(start, end), nextOffset: end };
}

async function postJson<T>(url: string, body: unknown): Promise<T> {
  const responseBody = await httpPostJson(
    url,
    "application/json",
    JSON.stringify(body),
  );
  try {
    return JSON.parse(responseBody.toString("utf8")) as T;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw PrayError.parse("auth response", message);
  }
}

function trimTrailingSlash(value: string): string {
  return value.replace(/\/+$/, "");
}
